### PR TITLE
Channel options

### LINKED ISSUE #120: Support party specific channel grpc_options
We are using grpc for cross party data sending, right now we have `fed._private.grpc_options` which is global, we need to support
party specific channel grpc_options, so each party could have their own options:

```
grpc.aio.insecure_channel("proxy", options=[('grpc.default_authority', 'alice')])
grpc.aio.insecure_channel("proxy", options=[('grpc.default_authority', 'bob')])
```

This is useful if differenct party server have different configs.

### PR DESCRIPTION
**Pull request by NKcqx**
This should close #120

### CODE REVIEW

File: fed/barriers.py [hunk 1]
```diff
@@ -230,36 +235,52 @@ async def send(
...
+            self._cluster[dest_party].get('grpc_metadata', {})
+        )
+        # merge grpc metadata
+        dest_party_grpc_config['grpc_metadata'] = {**global_grpc_metadata, **dest_party_grpc_metadata}
+
+        global_grpc_options = get_grpc_options(self.retry_policy)
+        dest_party_grpc_options = dict(
+            self._cluster[dest_party].get('grpc_options', [])
+        )
+        dest_party_grpc_config['grpc_options'] = {**global_grpc_options, **dest_party_grpc_options}
```
Comment (fengsp) [inline fed/barriers.py:272]: global_grpc_options is a list?

Comment (NKcqx) [reply fed/barriers.py]: Either is OK. 
* grpc repo's [example](https://github.com/grpc/grpc/blob/master/examples/python/helloworld/greeter_client_with_options.py) use list of tuples
* grpc [document](https://grpc.github.io/grpc/node/grpc.Channel.html#:~:text=A%20map%20of%20channel%20options%20that%20will%20be%20passed%20to%20the%20core.%20The%20available%20options%20are%20listed%20in%20this%20document.) say it's a map

I've changed to a dict to follow the above.

File: fed/barriers.py [hunk 2]
```diff
@@ -126,9 +126,14 @@ async def send_data_grpc(
     metadata=None,
     tls_config=None,
     retry_policy=None,
+    grpc_options=None
 ):
+    if grpc_options is None:
+        grpc_options = get_grpc_options(retry_policy=retry_policy)
+    else:
+        grpc_options = fed_utils.dict2tuple(grpc_options)
+
```
Comment (jovany-wang) [inline fed/barriers.py:135]: More pythonic?
```python
grpc_options = fed_utils.dict2tuple(grpc_options) if grpc_options is not None else get_grpc_options(retry_policy=retry_policy)
```

File: fed/barriers.py [hunk 3]
```diff
@@ -230,36 +235,54 @@ async def send(
...
                 retry_policy=self.retry_policy,
+                grpc_options=dest_party_grpc_config["grpc_options"]
             )
         except Exception as e:
             logger.error(f'Failed to {send_log_msg}, error: {e}')
             return False
         logger.debug(f"Succeeded to send {send_log_msg}. Response is {response}")
         return True  # True indicates it's sent successfully.
 
+    def setup_dest_party_grpc_config(self, dest_party):
```
Comment (jovany-wang) [inline fed/barriers.py:257]:
```suggestion
    def _setup_grpc_config(self, dest_party):
```

**Review by jovany-wang — APPROVED** [review]
It looks good to me. Just left 2 minor comments.

File: fed/cleanup.py [hunk 4]
```diff
@@ -45,6 +45,8 @@ def _signal_exit():
         os.kill(os.getpid(), signal.SIGTERM)
 
     global _sending_obj_refs_q
+    if not _sending_obj_refs_q:
```
[the remark this answers is not in the document]
Comment (NKcqx) [reply fed/cleanup.py:48]: Yes, here is the working PR #125

**Pull request merged**